### ISSUE TITLE
ligatures font support

### ISSUE THREAD

**Issue by Zane-XY** [neovim/neovim#1408]
It would be fantastic that NeoVim could support some ligatures programming font like https://github.com/i-tu/Hasklig.

There exists a patch to VIM  https://groups.google.com/forum/#!topic/vim_dev/0sETSAwe5Wo to enable this.

**Comment by Rydgel** [comment]
I would that too!

**Comment by lozandier** [comment]
Support for monotype fonts with ligatures would be a *fantastic* addition to this project. In addition to Hasklig and other monototype fonts such as [Pragmata Pro](http://www.fsd.it/fonts/pragmatapro.htm)

**Comment by sethwoodworth** [comment]
There are a few plugins that approximate ligatures using syntax files and the `conceal` feature.

In [vim-cute-python](https://github.com/ehamberg/vim-cute-python) it works like this:

```
syntax keyword pyNiceOperator sum conceal cchar=∑
syntax keyword pyNiceBuiltin all conceal cchar=∀
syntax keyword pyNiceBuiltin any conceal cchar=∃
```

While in a GUI vim, I would love to have a nuanced implementation of this feature (involving system clipboard taking the non-ligatured version of a glyph).  In terminal vim, I think I am happy with the builtin conceal.

**Comment by tsion** [comment]
Ligature fonts are much nicer than `conceal` because they don't change the width of the code, so alignment is always preserved. As for terminals, it's really up to the terminal itself to support ligature fonts, not vim, so this issue is only about gvim (if/when neovim gets it).

**Comment by lozandier** [comment]
@sethwoodworth Seconding what @tsion said, actual monospace typefaces w/ ligatures that have their full capabilities enabled by neovim supporting ligatures is significantly better than what `conceal` can ever hope to do. 

Examples of such typefaces are  Pragmata Pro & Hasklig support

**Comment by johntyree** [comment]
What does this actually add, feature wise? This seems like it breaks all the assumptions behind fixed-width fonts in the first place.

**Comment by tsion** [comment]
It adds support for rendering fonts with ligatures. (As pointed out above, it's only an issue for support in an eventual GUI version of Neovim.) Check out the link in the original post, https://github.com/i-tu/Hasklig, for an example font with ligatures.

It doesn't really break any assumptions about fixed-width fonts because a ligature of 2 characters will still be 2 characters wide, for example. (This is what we don't like about `conceal` - it *does* break those assumptions.)

**Comment by geohump** [comment]
No. Please don't. Its unneeded complication.

Adding the ability to handle ligatures breaks the one-to-one correspondence between Characters-displayed on the Screen, and the Characters actually in the line of text. This applies no matter what character encoding you use, including Unicode Glyphs.

Ligatures add nothing to the power and usefulness of Vim/Neo-vim and will only serve to make the code much harder to maintain as it will require adding non-fixed font handling to a system that was never intended or designed to handle it.

Take the standard 80x23 terminal display (or my standard 160x55 terminal, it doesn't matter what the size is)

All the line length calculations can be messed up by this because the character length of the line (as opposed to the display-length of the line) can now be different for every line of the display.

For example, the word "off", normally 3 character cells wide, is now only 2 character cells wide.

(a character cell is the amount of space a fixed font takes up on the display. Those of use who write code for screen editors like vim, think in Character cells )

**Comment by tsion** [comment]
Ligatures add more to the usefulness of neovim than `conceal` does, at least, in my opinion. They're another nice feature along the lines of syntax highlighting — sometimes very nice to have, but not strictly necessary.

It's not the end of the world to not have them, but it would be kind of sad if it would be so hard for neovim to support them. Even in the aged original vim codebase, it can be supported with a simple patch. (Bram simply didn't accept the patch because of typical fear of changing things in an old codebase, if I recall correctly.)

**Comment by tsion** [comment]
> All the line length calculations can be messed up by this because the character length of the line (as opposed to the display-length of the line) can now be different for every line of the display.
> 
> For example, the word "off", normally 3 character cells wide, is now only 2 character cells wide.

@geohump This isn't true. Ligatures don't make two character cells into one character cell - they just render a 2-character-cell-wide glyph over the two cells. The bytes stored in the file and in memory are still "o", "f", "f". (Also, people don't want this for ligatures like "ff", more like for "->" as in Hasklig.)

This is a font rendering thing.

**Comment by chrisbra** [comment]
On Fr, 30 Okt 2015, Scott Olson wrote:
> Even in the aged original vim codebase, it can be supported with a
> simple patch. (Bram simply didn't accept the patch because of typical
> fear of changing things in an old codebase, if I recall correctly.)

IIRC there was a simple patch aiming to make fond rendering work better, 
but it did not enable ligatures. AFAIK there is no patch for font 
ligature available for vim.

Best,
Christian

**Comment by ZyX-I** [comment]
> Adding the ability to handle ligatures breaks the one-to-one correspondence between Characters-displayed on the Screen, and the Characters actually in the line of text. This applies no matter what character encoding you use, including Unicode Glyphs.

There was never such a thing:

1. Tab character may take from one to &tabstop number of display cells.
2. Fullwidth characters take two display cells.
3. Combining characters take no display cells.
4. ASCII control characters take two display cells (`^X`) (or four, depends on some setting).
5. Unicode control characters, and also bytes that do not form a UTF-8 character in a UTF-8 file take from four to ten (if I am not mistaking) display cells (`<xx>`).

So there is no one-to-one correspondence that may be broken. If `ff` have taken one display cell in place of two it would though break alignment and things like block selection, but not what you say. Bun since they don’t break this I do not see any objections.

**Comment by geohump** [comment]
On Fri, Oct 30, 2015 at 7:46 PM, Scott Olson <notifications@github.com> wrote:
>
> All the line length calculations can be messed up by this because the character length of the line (as opposed to the display-length of the line) can now be different for every line of the display.
>
> For example, the word "off", normally 3 character cells wide, is now only 2 character cells wide.
>
> @geohump This isn't true. Ligatures don't make two character cells into one character cell - they just render a 2-character-cell-wide glyph over the two cells. The bytes stored in the file and in memory are still "o", "f", "f". (Also, people don't want this for ligatures like "ff", more like for "->" as in Hasklig.)
>
> This is a font rendering thing.
>

Well, I've only been doing this for 40 years now... :-)

The original reason ligatures existed was to save horizontal display space.

When the first printing presses came about, the fonts were cast in
metal and each letter occupied a different amount of space based on
the font. The work of "setting type" which became known as
"typesetting" was a load aureus hand worked job. You set the letters,
piece by piece into a frame to construct the sentences. Since paper
was expensive, and printing the paper was incredibly labor intensive
and time-consuming, you wanted to get as much text on each page as you
could. Ligatures were a way of doing that.

By reducing the amount of horizontal display space certain common
letter combinations used, you could increase the amountof text
contained on each page. As a newspaper publisher or book printer. This
save you money and increased your profits.

One additional benefit was a better looking line of text.

So the purpose of ligatures seem to be contradiction. Your claim that
ligatures are two character cells wide. That would completely defeat
their purpose. There's no reason for ligatures to exist except for
saving horizontal display space.

Now as for your glyph"->" taking two character cells on the display
screen, that is a glyph. That is a graphical character, not a
ligature.

I'm sorry that you have confused the two, but they do not serve the
same purpose. I understand why you confuse the two, because of the two
character cell mapping conflated with a two characters in one cell
mapping.

**Comment by geohump** [comment]
On Sat, Oct 31, 2015 at 8:45 AM, Nikolai Aleksandrovich Pavlov
<notifications@github.com> wrote:
>
> Adding the ability to handle ligatures breaks the one-to-one correspondence between Characters-displayed on the Screen, and the Characters actually in the line of text. This applies no matter what character encoding you use, including Unicode Glyphs.
>
> There was never such a thing:
>
> Tab character may take from one to &tabstop number of display cells.
> Fullwidth characters take two display cells.
> Combining characters take no display cells.
> ASCII control characters take two display cells (^X) (or four, depends on some setting).
> Unicode control characters, and also bytes that do not form a UTF-8 character in a UTF-8 file take from four to ten (if I am not mistaking) display cells (<xx>).
>
> So there is no one-to-one correspondence that may be broken. If ff have taken one display cell in place of two it would though break alignment and things like block selection, but not what you say. Bun since they don’t break this I do not see any objections.

Your information is valid and correct.

Unfortunately, it doesn't address the idea of ligatures.

Ligatures are specific instances of two letter combinations in a
specific font that take up only a single character cell of horizontal
display space.

Everything else you've mentioned here are lacking that characteristic.

If someone is talking about handling ligatures as a two character cell
wide display character, then why even bother? Just use those two
single character cell wide characters.

**Comment by tsion** [comment]
> When the first printing presses came about, the fonts were cast in
> metal and each letter occupied a different amount of space based on
> the font. The work of "setting type" which became known as
> "typesetting" was a load aureus hand worked job. You set the letters,
> piece by piece into a frame to construct the sentences. Since paper
> was expensive, and printing the paper was incredibly labor intensive
> and time-consuming, you wanted to get as much text on each page as you
> could. Ligatures were a way of doing that.
> 
> By reducing the amount of horizontal display space certain common
> letter combinations used, you could increase the amountof text
> contained on each page. As a newspaper publisher or book printer. This
> save you money and increased your profits.

@geohump Ok, thanks for the history lesson.

> One additional benefit was a better looking line of text.

This is literally the only reason this issue was opened.

> So the purpose of ligatures seem to be contradiction. Your claim that
> ligatures are two character cells wide. That would completely defeat
> their purpose. There's no reason for ligatures to exist except for
> saving horizontal display space.

Literally just above you gave a different reason for their purpose. Again I advise you to read https://github.com/i-tu/Hasklig#hasklig---ligatures-for-haskell-code and try to understand what this issue is actually asking for.

**Comment by kazuoteramoto** [comment]
@geohump Your idea of ligature is not the same as is used in this issue. As @tsion said reading the Hasklig wiki can help. In particular this part:

> The underlying code stays the same — only the representation changes.

In Hasklig and other monospaced fonts with ligatures, like Pragmata, the representation is space faithful, its uses the same horizontal space as the underling characters in the source code.

**Comment by rafaelrinaldi** [comment]
I would love to see font ligatures support on NeoVim :+1:

**Comment by agauniyal** [comment]
Jumping into the hipster bandwagon, eh..? Before implementing them, try to focus on other issues and PR's that are sitting there waiting to get resolved.

**Comment by tsion** [comment]
@agauniyal This issue is already blocked on even *having* a GUI interface for neovim in the first place. I'm sure the devs can figure out what to prioritize, anyway. No need to come around with insults or whatever it is you intended with the bandwagon comment.

**Comment by julienchurch** [comment]
@tsion  Landed on this thread kind of by accident and remembered how I'd always wanted to try Hasklig. Ever work anything out for yourself? 

(Or did you realize how wrong you were to test the ligature knowledge of someone with over 40 years' experience? Kidding, of course. :stuck_out_tongue_winking_eye: )

**Comment by tsion** [comment]
@julienchurch It's possible to make Hasklig work with gvim and the third party patch linked in the OP (https://groups.google.com/forum/#!topic/vim_dev/0sETSAwe5Wo) but I'm not aware of any other solutions for vim or neovim. I'm currently using vim primarily in a terminal these days, anyway, so ligature code fonts aren't really possible. At least, I haven't heard of any terminals supporting them.

**Comment by julienchurch** [comment]
@tsion Haha, that's pretty much sums my experience up after I wrote my comment. Looked for some info, tried it with gvim, then Atom, then finally just gave up the dream because I'll never use either of them. GUIs don't work for me. Staying within the terminal is just too convenient.

I appreciate the response though!

**Comment by brahm92** [comment]
+1 for ligature support for nvim gui. 
@geohump - I respect your clear expertise in the field. However, I disagree with you on one point - that ligature support doesn't add anything functionality wise. The fact that it looks nicer is, in my opinion, a significant advantage when it comes to a text editor. Personally, I would like such a support in nvim gui for the mere reason that it feels nicer to use. I am sure I am not alone, as is clear in this thread.

**Comment by romgrk** [comment]
Idk if it has been noted, but various terminals support ligatures. From what I remember Konsole does. (And so should any other Qt-made terminal). I have also seen JavaScript-HTML terminal emulators providing (maybe inadvertently) ligatures.

**Comment by james2doyle** [comment]
You can use ligatures in [rogual/neovim-dot-app](https://github.com/rogual/neovim-dot-app) right now actually.

![screen shot 2016-06-03 at 2 15 43 pm](https://cloud.githubusercontent.com/assets/1425304/15792958/b572e064-2995-11e6-926d-bf896208b846.png)

Looks hot!

**Comment by shofel** [comment]
It works for me in Konsole. The font is Fira Code
![ligatures-konsole](https://cloud.githubusercontent.com/assets/1355255/19239587/2576d3bc-8f20-11e6-98ce-cd3fd15bcabc.png)

**Comment by joeycozza** [comment]
Firacode works well with many different terminals.

[https://github.com/tonsky/FiraCode#editor-support](https://github.com/tonsky/FiraCode#editor-support)

These are the steps that I followed to make it work on Mac OS X with default Terminal (and of course neovim)

1. Download firacode from their github repo
2. Unzipped file
3. Opened "Font Book" application on Mac
4. Go to File -> Add Fonts
5. Choose the ttf folder from what you unzipped
6. In your terminal Go to Preferences, and choose FiraCode as your font.
7. Enjoy

**Comment by justinmk** [comment]
Oni frontend for Neovim supports ligatures: https://github.com/extr0py/oni/issues/138

Not sure if it addresses the issues that @equalsraf [mentioned](https://github.com/neovim/neovim/issues/1408#issuecomment-61958088), such as cursor movement and selection. cc @extr0py

**Comment by extr0py** [comment]
Cheers! Here's a quick demo of what it looks like with https://github.com/tonsky/FiraCode in ONI:

![ligature](https://cloud.githubusercontent.com/assets/13532591/22066558/48f932f6-dd42-11e6-8aa3-c14e724ee97c.gif)

It's purely a visual representation so, right now, cursor movement works as it would in the non-ligature case. One feature though that would be helpful would be to reveal the actual underlying character as the cursor moves on top of the ligature. Added this issue to track: https://github.com/extr0py/oni/issues/172

**Comment by justinmk** [comment]
Given that the original request here was to "support ligatures" it seems reasonably achieved. If there is a more specific request, a new issue should be opened.  ("Support ligatures in the terminal" isn't specific enough: say what exactly that means, what change is needed in nvim core.)

**Comment by solson** [comment]
@justinmk Agreed. Actually, it was never really an actionable issue for neovim, since it would only apply if there was an official GUI frontend in this repo. As far as I know, terminals shouldn't ever require neovim support to display ligatures.

**Comment by nomasprime** [comment]
Trying to use [Pragmata Pro # ligatures](https://www.fsd.it/2019/08/02/pragmatapro-0-828-runes-teletext-graphics-and-all-the-html5-entities/) with NeoVim. They don't work.

`[]` works in a Markdown file but not in a Ruby one.

Not sure if this is a NeoVim or iTerm issue so I've [raised there](https://gitlab.com/gnachman/iterm2/issues/7461) too.
